src/main/java/bio/terra/service/resourcemanagement/google/GoogleResourceService.java

```diff
@@ -572,6 +586,55 @@ public void enableIamPermissions(Map<String, List<String>> userPermissions, Stri
         throw new EnablePermissionsFailedException("Cannot enable iam permissions", lastException);
     }
 
+    public void removeIamPermissions(Map<String, List<String>> userPermissions, String projectId)
+        throws InterruptedException {
+
+        GetIamPolicyRequest getIamPolicyRequest = new GetIamPolicyRequest();
+
+        Exception lastException = null;
+        int retryWait = INITIAL_WAIT_SECONDS;
+        for (int i = 0; i < RETRIES; i++) {
```

Comment:
This should be refactored before merging, if it continues to match the code in `enableIamPermissions`

Reply:
It doesn't quite match.  I was on the fence but I think there's enough of a difference that it's a bit awkward to refactor

Reply:
How about
```
    
    private interface IamPermissionsFunc {
        void call(GetIamPolicyRequest getIamPolicyRequest) throws IOException, GeneralSecurityException;
    }
    
    private void iamPermissionsWithRetry(String operation, IamPermissionsFunc func) throws InterruptedException {

        GetIamPolicyRequest getIamPolicyRequest = new GetIamPolicyRequest();

        Exception lastException = null;
        int retryWait = INITIAL_WAIT_SECONDS;
        for (int i = 0; i < RETRIES; i++) {
            try {
                func.call(getIamPolicyRequest);
            } catch (IOException | GeneralSecurityException ex) {
                logger.info("Failed to " + operation + " iam permissions. Retry " + i + " of " + RETRIES, ex);
                lastException = ex;
            }

            TimeUnit.SECONDS.sleep(retryWait);
            retryWait = retryWait + retryWait;
            if (retryWait > MAX_WAIT_SECONDS) {
                retryWait = MAX_WAIT_SECONDS;
            }
        }
        throw new EnablePermissionsFailedException("Cannot " + operation + " iam permissions", lastException);
    }

    public void enableIamPermissions(Map<String, List<String>> userPermissions, String projectId)
        throws InterruptedException {
        
        iamPermissionsWithRetry("enable", getIamPolicyRequest -> {
            CloudResourceManager resourceManager = cloudResourceManager();
            Policy policy = resourceManager.projects()
                .getIamPolicy(projectId, getIamPolicyRequest).execute();
            List<Binding> bindingsList = policy.getBindings();

            for (Map.Entry<String, List<String>> entry : userPermissions.entrySet()) {
                Binding binding = new Binding()
                    .setRole(entry.getKey())
                    .setMembers(entry.getValue());
                bindingsList.add(binding);
            }

            policy.setBindings(bindingsList);
            SetIamPolicyRequest setIamPolicyRequest = new SetIamPolicyRequest().setPolicy(policy);
            resourceManager.projects()
                .setIamPolicy(projectId, setIamPolicyRequest).execute();
        });
    }
```

Reply:
Yeah, but you'd also need to pass in the exception class to throw and that feels a little odd

Reply:
well, this compiles at least
```

    private void iamPermissionsWithRetry(String operation, IamPermissionsFunc func, 
                                         BiFunction<String, Exception, InternalServerErrorException> whenFail) throws InterruptedException {

        GetIamPolicyRequest getIamPolicyRequest = new GetIamPolicyRequest();

        Exception lastException = null;
        int retryWait = INITIAL_WAIT_SECONDS;
        for (int i = 0; i < RETRIES; i++) {
            try {
                func.call(getIamPolicyRequest);
                break;
            } catch (IOException | GeneralSecurityException ex) {
                logger.info("Failed to " + operation + " iam permissions. Retry " + i + " of " + RETRIES, ex);
                lastException = ex;
            }

            TimeUnit.SECONDS.sleep(retryWait);
            retryWait = retryWait + retryWait;
            if (retryWait > MAX_WAIT_SECONDS) {
                retryWait = MAX_WAIT_SECONDS;
            }
        }
        throw whenFail.apply("Cannot " + operation + " iam permissions", lastException);
    }

    public void enableIamPermissions(Map<String, List<String>> userPermissions, String projectId)
        throws InterruptedException {

        iamPermissionsWithRetry("enable", getIamPolicyRequest -> {
            CloudResourceManager resourceManager = cloudResourceManager();
            Policy policy = resourceManager.projects()
                .getIamPolicy(projectId, getIamPolicyRequest).execute();
            List<Binding> bindingsList = policy.getBindings();

            for (Map.Entry<String, List<String>> entry : userPermissions.entrySet()) {
                Binding binding = new Binding()
                    .setRole(entry.getKey())
                    .setMembers(entry.getValue());
                bindingsList.add(binding);
            }

            policy.setBindings(bindingsList);
            SetIamPolicyRequest setIamPolicyRequest = new SetIamPolicyRequest().setPolicy(policy);
            resourceManager.projects()
                .setIamPolicy(projectId, setIamPolicyRequest).execute();
        }, EnablePermissionsFailedException::new);
    }
```

Reply:
Right...this was why I was saying I think it gets a little awkward

Reply:
Readability is a subjective measure. I'm not saying the code I posted is more readable, but I think we can agree that there is a separation of concerns issue with this code as currently written (in addition to the code duplication). One concern is retrying, the other is the operation itself. The two are independent but currently this is not expressed in the code, so it must be re-derived each time by the reader.

Reply:
Do we have a shared method to do simple retries like this?  If not, should we?